### PR TITLE
Introduce internal API call queuing mechanism

### PR DESCRIPTION
**Pull request by laurynas-karvelis**
Introduce internal API call queuing mechanism to send messages sequentially (no more random order)

### CODE REVIEW

**Comment by laurynas-karvelis** [comment]
Thank you for the merge of my previous PR and super quick turnaround :).
This PR will be probably the most wanted feature from this module for my BE app :). Queuing! :clap:

File: src/functions/api.js [hunk 1]
```diff
@@ -24,12 +59,24 @@ API.prototype.request = function request(method, data) {
...
+    // implementation taken from https://github.com/yagop/node-telegram-bot-api/issues/192#issuecomment-249488807
+    let resolve;
+    let reject;
+
+    const promise = new Promise((_resolve, _reject) => {
+      resolve = _resolve;
+      reject = _reject;
+    });
+
+    this._queue.push({ method, data, resolve, reject });
```
Comment (mdibaiee) [inline src/functions/api.js]: Why not wrap this function inside the promise, so you don't have to hold references to `resolve` and `reject`.

```javascript

return new Promise((resolve, reject) => {
  this._queue.push({ method, data, resolve, reject });
  process.nextTick(this._runQueue.bind(this));
});
```

**Review by mdibaiee — CHANGES_REQUESTED** [review]
Thanks! This is definitely a welcome improvement, there is a single detail I've commented below

File: src/functions/api.js (hunk 1 above)
Comment (laurynas-karvelis) [reply src/functions/api.js]: You are right, after re-running same test file that sends out multiple messages got same result without any issues. Appears there's no negative impact reducing this code excerpt you suggested :+1: 
Added your suggestions in a new commit

**Comment by mdibaiee** [comment]
Looks good to me, just to make sure, have you tested this thoroughly? Asking since this can have implications on most parts of the application. If so, good to go by me, thanks a lot! :D

**Comment by laurynas-karvelis** [comment]
I have tested it out with simple Messages, no document uploads etc. But from the tests I've done, it worked like a charm for me.

**Comment by mdibaiee** [comment]
Could you please put some time to test other types of messages just to make sure? I would be grateful :pray:

**Comment by laurynas-karvelis** [comment]
As every API method is just a wrapper for API.request method, and param count to it remains the same, I see no reason other child classes of Base should behave differently. I haven't dealt with sending docs etc via Telegram API before. Could be a possibility for you to test this out quicker than me by any chance? :)

**Comment by mdibaiee** [comment]
I guess you are right, since the change only affects the way the method is _called_ rather than processed, it shouldn't make any difference among different types of messages.

Thanks a lot Laurynas! :+1:

**Comment by laurynas-karvelis** [comment]
:tada:

**Comment by mdibaiee** [comment]
Published `telegram-api@4.0.0` :tada::+1:

**Pull request merged**